pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -230,3 +234,103 @@ func buildGatewayInboundHTTPRouteConfig(env model.Environment, gatewayName strin
 		VirtualHosts: virtualHosts,
 	}
 }
+
+// buildGatewayInboundNetworkFilters retrieves all VirtualServices bound to the set of Gateways for this workload, filters
+// them by this server's port and hostnames, and produces network filters for each destination from the filtered services
+func buildGatewayNetworkFilters(env model.Environment, server *networking.Server, gatewayNames []string) []listener.Filter {
+	port := &model.Port{
+		Name:     server.Port.Name,
+		Port:     int(server.Port.Number),
+		Protocol: model.ConvertCaseInsensitiveStringToProtocol(server.Port.Protocol),
+	}
+
+	dests := filterTCPDownstreams(env, server, gatewayNames)
+	filters := make([]listener.Filter, 0, len(dests))
+	for _, dest := range dests {
+		filters = append(filters, buildOutboundNetworkFilters(destToClusterName(dest), []string{dest.Host}, port)...)
+	}
+	return filters
+}
+
+// filterTCPDownstreams filters virtual services by gateway names, then determines if any match the (TCP) server
+// TODO: move up to more general location so this can be re-used in sidecars
+func filterTCPDownstreams(env model.Environment, server *networking.Server, gatewayNames []string) []*networking.Destination {
+	hosts := make(map[string]bool, len(server.Hosts))
+	for _, host := range server.Hosts {
+		hosts[host] = true
+	}
+
+	gateways := make(map[string]bool, len(gatewayNames))
+	for _, gateway := range gatewayNames {
+		gateways[gateway] = true
+	}
+
+	virtualServices := env.VirtualServices(gatewayNames)
+	downstreams := make([]*networking.Destination, 0, len(virtualServices))
+	for _, spec := range virtualServices {
+		vsvc := spec.Spec.(*networking.VirtualService)
+
+		// TODO: real wildcard based matching; does code to do that not exist already?
+		match := false
+		for _, host := range vsvc.Hosts {
+			match = match || hosts[host]
+		}
+		if !match {
+			// the VirtualService's hosts don't include hosts advertised by server
+			continue
+		}
+
+		// hosts match, now we ensure we satisfy the rule's l4 match conditions, if any exist
+		for _, tcp := range vsvc.Tcp {
+			if l4Match(tcp.Match, server, gateways) {
+				downstreams = append(downstreams, gatherDestinations(tcp.Route)...)
+			}
+		}
+	}
+	return downstreams
+}
+
+// TODO: move up to more general location so this can be re-used in other service matching
+func l4Match(predicates []*networking.L4MatchAttributes, server *networking.Server, gatewayNames map[string]bool) bool {
```

Comment:
nit: add a TODO here saying we need to do more match stuff, like honoring cidr prefix matches

Reply:
added below inside the loop